File: solutions/ruby/allergies/1/allergies.rb
class Allergies 
  ALLERGENS = {
    128 => "cats",
    64 => "pollen",
    32 => "chocolate", 
    16 => "tomatoes", 
    8 => "strawberries",
    4 => "shellfish", 
    2 => "peanuts", 
    1 => "eggs"
  }.freeze 
  
  def initialize(score)
    @score = score 
  end 

  def allergic_to?(allergen)
    list.include?(allergen)
  end 

  def list
    ALLERGENS.map do |score, allergy|
      allergic, @score = @score.divmod(score)
      allergy if allergic % 2 == 1 
    end.compact.reverse 
  end 
end 

File: solutions/cpp/allergies/1/allergies.cpp
#include "allergies.h"

namespace allergies {
    Allergy::Allergy(int score) : score(score) {
        for(const auto& pair : ALLERGIES)
            if ((score & pair.second) == pair.second) 
                allergies.emplace(pair.first);
    }
    
    bool Allergy::is_allergic_to(const string& allergen) {
        return allergies.count(allergen);
    }
    
    unordered_set<string>  Allergy::get_allergies() {
        return allergies; 
    }
    
    Allergy allergy_test(int score) {
        return Allergy{score};
    }
}  // namespace allergies
